### PR TITLE
fix(behavior_path_planner): fix abnormal emergency stop when enable_avoidance_over_opposite_direction is true

### PR DESCRIPTION
**Pull request by shulanbushangshu**
## Description
When the parameter "enable_avoidance_over_opposite_direction"is true(The default is true), the vehicle uses the reverse lane for obstacle avoidance; When the vehicle avoids the obstacles and returns to the main lane, the vehicle will stop abnormally and then run again (because the calculated drivable area changes suddenly).
By this PR, it can fix the abnormal emergency stop in obstacle avoidance.

Before this PR:
https://user-images.githubusercontent.com/102840938/183551281-062b1912-62a9-4287-971f-4e6be5b0c5ea.mp4
After this PR：
https://user-images.githubusercontent.com/102840938/183551431-8e3e4335-e2d3-42de-8e33-5603f9ed3ef1.mp4

## Tests performed
Current simulation：

https://user-images.githubusercontent.com/102840938/183551281-062b1912-62a9-4287-971f-4e6be5b0c5ea.mp4

Modified simulation：

https://user-images.githubusercontent.com/102840938/183551431-8e3e4335-e2d3-42de-8e33-5603f9ed3ef1.mp4

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by takayuki5168** [comment]
@zulfaqar-azmi-t4 @TakaHoribe Could you review this.

**Comment by shulanbushangshu** [comment]
@zulfaqar-azmi-t4 ,could you check this simulated phenomenon?

**Comment by zulfaqar-azmi-t4** [comment]
@shulanbushangshu 
Hi, sorry for the late response.
Just to update, currently I am testing your solution.
I'll let your know tomorrow after I analyze your code a bit.

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 1]
```diff
@@ -1723,6 +1723,31 @@ void AvoidanceModule::generateExtendedDrivableArea(ShiftedPath * shifted_path) c
...
+        if (fabs(current_ego_shift) > 0.01) {
+          if (current_ego_shift > 0.0) {
+            search_lanelets = route_handler->getAllSharedLineStringLanelets(
+              current_lane, !get_right, get_left, include_opposite);
+          } else {
+            search_lanelets = route_handler->getAllSharedLineStringLanelets(
+              current_lane, get_right, !get_left, include_opposite);
+          }
+          if (search_lanelets.size() > 1) {
+            extended_lanelets.push_back(search_lanelets.at(1));
```
Comment (zulfaqar-azmi-t4) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:1744]: @shulanbushangshu 
I don't have a map to test, but technically, there is a possibility that the drivable area extension might will result with the same issue if the number of opposite lane > 1. 

Another thing of my concern in that multiple similar lanelet will be pushed back (i think the same issue already existed), and it might affect the performance a bit. 

I have another solution to this, but I will need to discuss internally about this part, and let you know about it.

Comment (shulanbushangshu) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:1744]: Thank you for your examination.I  use "extended_lanelets.push_back(search_lanelets.at(1)) " to ensure that the current adjacent lane is used instead of multiple lanes.I have tested the two opposite lanes and the result is well.


https://user-images.githubusercontent.com/102840938/184335777-0429403f-4411-4dc5-b807-f8e61b601d02.mp4

**Comment by zulfaqar-azmi-t4** [comment]
Hi @shulanbushangshu 
We have discussed internally for the workaround of this.

From the look of it, the drivable area extension is too conservative, which add complexity to the current method. 
One simpler solution is to just extend the drivable area to both left and right of `avoidance_data_.current_lanelets`. This way, 
1. we can reduce the complexity by not having conditional statement.
2. we can remove [this](https://github.com/autowarefoundation/autoware.universe/blob/7566a66031fdffd7e3ba8e99ce948ce37efc793e/planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp#:~:text=avoidance_data_.current_lanelets%3B-,%7B,%7D,-%7B)
We don't really have to worry about the portion of the drivable area that is not used.

The portion of the code that should be modify is [here](https://github.com/autowarefoundation/autoware.universe/blob/7566a66031fdffd7e3ba8e99ce948ce37efc793e/planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp#:~:text=avoidance_data_.current_lanelets\)%20%7B-,%7B%20%20//%201.%20extend%20to%20right/left%20or%20adjacent%20right/left,%7D,-%7B%20%20//%202).

Basically is to have
```
        search_lanelets = route_handler->getAllSharedLineStringLanelets(
          lane);
      }
      extended_lanelets.insert(
        extended_lanelets.end(), search_lanelets.begin(), search_lanelets.end());
```

Let me know what do you think about this.

**Comment by shulanbushangshu** [comment]
Hi,@zulfaqar-azmi-t4
Yes, the original idea of considering the driveable area which would be used leds to the complexity of my previous code.
As you think, we can simply add the left and right lanes of the current lane to the extension Lanes.
I change the code as you say and simulate:
Screenshot from 2022-08-15 13-13-05

case1:
https://user-images.githubusercontent.com/102840938/184581248-431ed969-50b4-492b-aeb8-95984435a4c0.mp4
case2:
https://user-images.githubusercontent.com/102840938/184582823-ceb69f91-4d42-414d-9028-ab083bcd86ed.mp4

The result is well.The drivable area will remain the maximum range during the operation of the obstacle avoidance module.
By the way, whether or not the maximum driveable area increases the amount of calculation in the module"obstacle_avoidance_planner"

**Comment by zulfaqar-azmi-t4** [comment]
Hi, @shulanbushangshu. Thank you very much for the confirmation.

Great. Considering `enable_avoidance_over_opposite_direction`, a little more stuff might need to be added.

About `obstacle_avoidance_planner`, I have checked with the corresponding developer about the impact of the changes and it seems to be fine.

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp (hunk 1 above)
Comment (taikitanaka3) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:1744]: @shulanbushangshu 
@zulfaqar-azmi-t4 -san says search_lanelets can be size one and that will cause segmentation fault at `extended_lanelets.push_back(search_lanelets.at(1))`

I think the point for this PR is getCurrentShift() can be more suitable than isOnRight() as you show.
@shulanbushangshu can you please consider that reason at first?

**Comment by zulfaqar-azmi-t4** [comment]
Hi @shulanbushangshu
Just checking in case you need some help regarding this. If yes, then I can make some changes from my branch and make a separate PR.

**Comment by zulfaqar-azmi-t4** [comment]
Hi @shulanbushangshu 

Due to some internal discussion and request, we also requires this part to be fix. Therefore, I deeply apologize for having to create another PR to fix this part. The PR included the details of what we have discussed, with some additional refactoring. 

https://github.com/autowarefoundation/autoware.universe/pull/1688

**Comment by shulanbushangshu** [comment]
@zulfaqar-azmi-t4
Yes,I have read the new code.It works well.I will close this pr.
